share/util/source_stat.py

```diff
@@ -0,0 +1,158 @@
+import logging
+
+import requests
+from furl import furl
+from lxml import etree
+import pendulum
+
+from share.models import SourceStat
+from share.models import SourceConfig
+
+logger = logging.getLogger(__name__)
+
+
+class SourceStatus():
+
+    def assert_no_exception(self, url, timeout=15.0):
+        try:
+            r = requests.get(url, timeout=timeout)
+        # except all exception and log
+        except Exception as e:
+            logger.warning('Exception received from source: %s', e)
+            return (None, e)
+        return (r, None)
+
+    def get_source_stats(self, config_id):
+        sourceconfig = SourceConfig.objects.get(pk=config_id)
+
+        base_url_config = sourceconfig.base_url
+        response_elapsed_time = 0
+        response_status_code = None
+
+        response, response_exception = self.assert_no_exception(base_url_config)
+
+        if response is not None:
+            response_elapsed_time = response.elapsed.total_seconds()
+            response_status_code = response.status_code
+
+        self.create_source_stat(
+            config_id,
+            earliest_datestamp_config=str(sourceconfig.earliest_date) if sourceconfig.earliest_date else None,
+            base_url_config=base_url_config,
+            response_status_code=response_status_code,
+            response_elapsed_time=response_elapsed_time,
+            response_exception=response_exception,
+        )
+
+    def create_source_stat(self, config_id, earliest_datestamp_source=None,
+                           earliest_datestamps_match=True, base_url_source='',
+                           base_urls_match=True, admin_note='', **kwargs):
+        SourceStat.objects.create(
+            earliest_datestamp_source=earliest_datestamp_source,
+            earliest_datestamp_config=kwargs.pop('earliest_datestamp_config'),
+            earliest_datestamps_match=earliest_datestamps_match,
+
+            base_url_source=base_url_source,
+            base_url_config=kwargs.pop('base_url_config'),
+            base_urls_match=base_urls_match,
+
+            config_id=config_id,
+
+            response_status_code=kwargs.pop('response_status_code'),
+            response_elapsed_time=kwargs.pop('response_elapsed_time'),
+            response_exception=kwargs.pop('response_exception'),
+
+            admin_note=admin_note,
+        )
+
+
+class OAISourceStatus(SourceStatus):
+
+    NAMESPACES = {
+        'dc': 'http://purl.org/dc/elements/1.1/',
+        'ns0': 'http://www.openarchives.org/OAI/2.0/',
+        'oai_dc': 'http://www.openarchives.org/OAI/2.0/oai_dc/',
+    }
+
+    def get_field_from_identify(self, response, field):
+        parsed = etree.fromstring(response.content, parser=etree.XMLParser(recover=True))
```

Comment:
Not sure if there's a great way to do this, might just be worth adding a todo. It would be interesting to see what sources provide invalid XML. IE fail without `recover=True`